File: mobile/lib/features/pagos/data/pago_processor.dart
import '../../../data/app_database.dart';

export '../../../data/app_database.dart' show Cuota, Pago, Prestamo;

/// El abono supera lo pendiente de la cuota y hace falta que el cobrador
/// indique cómo manejar el excedente antes de poder calcular el plan de pago.
class ManejoExcedenteRequeridoException implements Exception {
  ManejoExcedenteRequeridoException(this.excedente);

  final double excedente;
}

/// El abono no alcanza a cubrir lo pendiente de la cuota y hace falta que el
/// cobrador elija qué política de mora aplicar antes de poder calcular el
/// plan de pago.
class PoliticaMoraRequeridaException implements Exception {
  PoliticaMoraRequeridaException(this.faltante);

  final double faltante;
}

/// Una fila que debe insertarse en `pagos`.
class PagoAInsertar {
  const PagoAInsertar({
    required this.cuotaId,
    required this.montoAbonado,
    required this.montoAplicado,
    required this.fechaPago,
    required this.diasMora,
    required this.saldoRestanteDespues,
  });

  final int cuotaId;
  final double montoAbonado;
  final double montoAplicado;
  final DateTime fechaPago;
  final int diasMora;
  final double saldoRestanteDespues;
}

/// Un cambio que debe aplicarse a una cuota existente como resultado del pago.
class ActualizacionCuota {
  const ActualizacionCuota({required this.cuotaId, this.nuevoEstado, this.nuevoMontoEsperado});

  final int cuotaId;
  final String? nuevoEstado;
  final double? nuevoMontoEsperado;

  @override
  bool operator ==(Object other) =>
      other is ActualizacionCuota &&
      other.cuotaId == cuotaId &&
      other.nuevoEstado == nuevoEstado &&
      other.nuevoMontoEsperado == nuevoMontoEsperado;

  @override
  int get hashCode => Object.hash(cuotaId, nuevoEstado, nuevoMontoEsperado);

  @override
  String toString() =>
      'ActualizacionCuota(cuotaId: $cuotaId, nuevoEstado: $nuevoEstado, nuevoMontoEsperado: $nuevoMontoEsperado)';
}

/// Resultado completo de procesar un pago: qué filas insertar en `pagos`, qué
/// cuotas actualizar, y el nuevo estado del préstamo. `politicaMoraAplicada`
/// solo viene informada cuando el abono fue insuficiente (faltante) y se usó
/// para decidir qué hacer con esa diferencia.
class PlanPago {
  const PlanPago({
    required this.pagos,
    required this.actualizacionesCuotas,
    required this.nuevoEstadoPrestamo,
    this.politicaMoraAplicada,
  });

  final List<PagoAInsertar> pagos;
  final List<ActualizacionCuota> actualizacionesCuotas;
  final String nuevoEstadoPrestamo;
  final String? politicaMoraAplicada;
}

/// Réplica en Dart de `App\Services\PagoProcessor` del backend Laravel: mismo
/// cálculo de mora, mismos tres escenarios (exacto/faltante/excedente),
/// mismas políticas de mora y mismo manejo de excedente. Cualquier cambio a
/// esta lógica en el backend debe reflejarse aquí también (ver nota en
/// CLAUDE.md sobre `PrestamoCalculator`, el mismo principio aplica).
///
/// A diferencia del backend (donde `politica_mora` es fija por préstamo), acá
/// se recibe como parámetro en cada llamada: la app permite que el cobrador
/// elija la política en el momento del pago cuando hay un faltante. Es
/// responsabilidad de quien use este resultado (`PagosRepository`) decidir si
/// esa elección también actualiza el préstamo para futuros pagos.
class PagoProcessor {
  const PagoProcessor();

  static const _politicasValidas = {'mantener', 'siguiente_pago', 'sumar_total'};
  static const _manejosExcedenteValidos = {'abono_deuda', 'cobro_extra'};

  /// Calcula el plan de pago. No toca la base de datos: [cuotas] y
  /// [pagosExistentes] deben reflejar el estado actual del préstamo, y
  /// [montoTotalPrestamo] es `PrestamoDetalle.montoTotal` (capital + interés
  /// + extras).
  ///
  /// Lanza [ManejoExcedenteRequeridoException] o [PoliticaMoraRequeridaException]
  /// si falta una decisión del cobrador para poder continuar; quien llame debe
  /// volver a invocar con el parámetro resuelto.
  PlanPago procesar({
    required List<Cuota> cuotas,
    required List<Pago> pagosExistentes,
    required double montoTotalPrestamo,
    required double montoAbonado,
    required DateTime fechaPago,
    String? politicaMora,
    String? manejoExcedente,
  }) {
    final pendientes = [...cuotas.where((c) => c.estado != 'pagada')]
      ..sort((a, b) => a.numeroCuota.compareTo(b.numeroCuota));
    if (pendientes.isEmpty) {
      throw StateError('Este préstamo no tiene cuotas pendientes por pagar.');
    }
    final cuotaActual = pendientes.first;

    final fechaEsperada = _soloFecha(cuotaActual.fechaEsperada);
    final fechaDePago = _soloFecha(fechaPago);
    final diasMora = fechaDePago.isAfter(fechaEsperada) ? fechaDePago.difference(fechaEsperada).inDays : 0;

    final pendienteEnCuota = _pendienteEnCuota(cuotaActual, pagosExistentes);
    final montoAbonadoRedondeado = _redondear(montoAbonado);

    final pagos = <PagoAInsertar>[];
    final actualizaciones = <ActualizacionCuota>[];
    String? politicaMoraAplicada;

    var totalAplicadoAcumulado = _redondear(
      pagosExistentes.fold<double>(0, (acumulado, pago) => acumulado + pago.montoAplicado),
    );

    PagoAInsertar crearPago(int cuotaId, double montoAbonadoFila, double montoAplicado, int diasMoraFila) {
      totalAplicadoAcumulado = _redondear(totalAplicadoAcumulado + montoAplicado);
      final saldoRestante = _redondear(montoTotalPrestamo - totalAplicadoAcumulado);
      return PagoAInsertar(
        cuotaId: cuotaId,
        montoAbonado: montoAbonadoFila,
        montoAplicado: montoAplicado,
        fechaPago: fechaPago,
        diasMora: diasMoraFila,
        saldoRestanteDespues: saldoRestante < 0 ? 0 : saldoRestante,
      );
    }

    if (montoAbonadoRedondeado > pendienteEnCuota) {
      if (manejoExcedente == null || !_manejosExcedenteValidos.contains(manejoExcedente)) {
        throw ManejoExcedenteRequeridoException(_redondear(montoAbonadoRedondeado - pendienteEnCuota));
      }

      final excedente = _redondear(montoAbonadoRedondeado - pendienteEnCuota);
      actualizaciones.add(ActualizacionCuota(cuotaId: cuotaActual.id, nuevoEstado: 'pagada'));

      final montoAbonadoFila = manejoExcedente == 'cobro_extra' ? montoAbonadoRedondeado : pendienteEnCuota;
      pagos.add(crearPago(cuotaActual.id, montoAbonadoFila, pendienteEnCuota, diasMora));

      if (manejoExcedente == 'abono_deuda') {
        _aplicarExcedenteComoAbono(
          cuotas: cuotas,
          cuotaActual: cuotaActual,
          pagosExistentes: pagosExistentes,
          excedenteInicial: excedente,
          crearPago: crearPago,
          pagos: pagos,
          actualizaciones: actualizaciones,
        );
      }
    } else if (montoAbonadoRedondeado < pendienteEnCuota) {
      if (politicaMora == null || !_politicasValidas.contains(politicaMora)) {
        throw PoliticaMoraRequeridaException(_redondear(pendienteEnCuota - montoAbonadoRedondeado));
      }

      final faltante = _redondear(pendienteEnCuota - montoAbonadoRedondeado);
      politicaMoraAplicada = politicaMora;

      _aplicarPoliticaMora(
        cuotas: cuotas,
        cuotaActual: cuotaActual,
        politica: politicaMora,
        faltante: faltante,
        diasMora: diasMora,
        actualizaciones: actualizaciones,
      );

      pagos.add(crearPago(cuotaActual.id, montoAbonadoRedondeado, montoAbonadoRedondeado, diasMora));
    } else {
      actualizaciones.add(ActualizacionCuota(cuotaId: cuotaActual.id, nuevoEstado: 'pagada'));
      pagos.add(crearPago(cuotaActual.id, montoAbonadoRedondeado, montoAbonadoRedondeado, diasMora));
    }

    final nuevoEstadoPrestamo = _calcularEstadoPrestamo(cuotas, actualizaciones);

    return PlanPago(
      pagos: pagos,
      actualizacionesCuotas: actualizaciones,
      nuevoEstadoPrestamo: nuevoEstadoPrestamo,
      politicaMoraAplicada: politicaMoraAplicada,
    );
  }

  double _pendienteEnCuota(Cuota cuota, List<Pago> pagosExistentes) {
    final yaAplicado = pagosExistentes
        .where((p) => p.cuotaId == cuota.id)
        .fold<double>(0, (acumulado, pago) => acumulado + pago.montoAplicado);
    return _redondear(cuota.montoEsperado - yaAplicado);
  }

  void _aplicarExcedenteComoAbono({
    required List<Cuota> cuotas,
    required Cuota cuotaActual,
    required List<Pago> pagosExistentes,
    required double excedenteInicial,
    required PagoAInsertar Function(int cuotaId, double montoAbonadoFila, double montoAplicado, int diasMoraFila)
    crearPago,
    required List<PagoAInsertar> pagos,
    required List<ActualizacionCuota> actualizaciones,
  }) {
    var excedente = excedenteInicial;

    final siguientes = [...cuotas.where((c) => c.estado != 'pagada' && c.numeroCuota > cuotaActual.numeroCuota)]
      ..sort((a, b) => a.numeroCuota.compareTo(b.numeroCuota));

    for (final siguiente in siguientes) {
      if (excedente <= 0) break;

      final pendiente = _pendienteEnCuota(siguiente, pagosExistentes);
      final aAplicar = excedente < pendiente ? excedente : pendiente;

      if (aAplicar >= pendiente) {
        actualizaciones.add(ActualizacionCuota(cuotaId: siguiente.id, nuevoEstado: 'pagada'));
      }

      pagos.add(crearPago(siguiente.id, aAplicar, aAplicar, 0));
      excedente = _redondear(excedente - aAplicar);
    }
  }

  void _aplicarPoliticaMora({
    required List<Cuota> cuotas,
    required Cuota cuotaActual,
    required String politica,
    required double faltante,
    required int diasMora,
    required List<ActualizacionCuota> actualizaciones,
  }) {
    var politicaEfectiva = politica;

    if (politicaEfectiva == 'siguiente_pago') {
      final siguiente = [...cuotas.where((c) => c.estado != 'pagada' && c.numeroCuota > cuotaActual.numeroCuota)]
          .fold<Cuota?>(null, (menor, c) => menor == null || c.numeroCuota < menor.numeroCuota ? c : menor);

      if (siguiente != null) {
        actualizaciones.add(
          ActualizacionCuota(cuotaId: siguiente.id, nuevoMontoEsperado: _redondear(siguiente.montoEsperado + faltante)),
        );
        actualizaciones.add(ActualizacionCuota(cuotaId: cuotaActual.id, nuevoEstado: 'pagada'));
        return;
      }

      politicaEfectiva = 'mantener';
    }

    if (politicaEfectiva == 'sumar_total') {
      final pendientes = [...cuotas.where((c) => c.estado != 'pagada' && c.id != cuotaActual.id)]
        ..sort((a, b) => a.numeroCuota.compareTo(b.numeroCuota));

      if (pendientes.isNotEmpty) {
        final porCuota = _redondear(faltante / pendientes.length);
        var acumulado = 0.0;

        for (var indice = 0; indice < pendientes.length; indice++) {
          final esUltima = indice == pendientes.length - 1;
          final incremento = esUltima ? _redondear(faltante - acumulado) : porCuota;
          acumulado = _redondear(acumulado + incremento);

          actualizaciones.add(
            ActualizacionCuota(
              cuotaId: pendientes[indice].id,
              nuevoMontoEsperado: _redondear(pendientes[indice].montoEsperado + incremento),
            ),
          );
        }

        actualizaciones.add(ActualizacionCuota(cuotaId: cuotaActual.id, nuevoEstado: 'pagada'));
        return;
      }
    }

    actualizaciones.add(
      ActualizacionCuota(cuotaId: cuotaActual.id, nuevoEstado: diasMora > 0 ? 'en_mora' : 'pendiente'),
    );
  }

  String _calcularEstadoPrestamo(List<Cuota> cuotas, List<ActualizacionCuota> actualizaciones) {
    String estadoDe(Cuota cuota) {
      final actualizacion = actualizaciones.lastWhere(
        (a) => a.cuotaId == cuota.id && a.nuevoEstado != null,
        orElse: () => ActualizacionCuota(cuotaId: cuota.id),
      );
      return actualizacion.nuevoEstado ?? cuota.estado;
    }

    final quedaPendiente = cuotas.any((c) => estadoDe(c) != 'pagada');
    if (!quedaPendiente) return 'pagado';

    final hayMora = cuotas.any((c) => estadoDe(c) == 'en_mora');
    return hayMora ? 'en_mora' : 'activo';
  }

  DateTime _soloFecha(DateTime fecha) => DateTime(fecha.year, fecha.month, fecha.day);

  double _redondear(double valor) => (valor * 100).round() / 100;
}

File: backend/app/Services/PagoProcessor.php
<?php

namespace App\Services;

use App\Models\Cuota;
use App\Models\Pago;
use App\Models\Prestamo;
use Illuminate\Support\Carbon;
use Illuminate\Support\Collection;
use Illuminate\Support\Facades\DB;
use RuntimeException;

class PagoProcessor
{
    /**
     * Registra un pago contra la cuota pendiente más antigua de un préstamo,
     * aplicando la política de mora del préstamo si el abono no la cubre por completo,
     * o pidiendo una decisión explícita si el abono la supera.
     *
     * Devuelve una colección de Pago: normalmente uno solo, pero puede haber varios
     * cuando un excedente (manejo_excedente=abono_deuda) alcanza a cubrir cuotas
     * futuras, ya que cada pago solo puede referenciar una cuota (cuota_id).
     *
     * @param  array{monto_abonado: float|string, fecha_pago: string, manejo_excedente?: string|null}  $datos
     * @return Collection<int, Pago>
     */
    public function procesar(Prestamo $prestamo, array $datos): Collection
    {
        return DB::transaction(function () use ($prestamo, $datos) {
            $prestamo = Prestamo::whereKey($prestamo->id)->lockForUpdate()->firstOrFail();

            if ($prestamo->estado === 'anulado') {
                throw new RuntimeException('No se pueden registrar pagos sobre un préstamo anulado.');
            }

            $cuota = $prestamo->cuotas()
                ->where('estado', '!=', 'pagada')
                ->orderBy('numero_cuota')
                ->lockForUpdate()
                ->first();

            if (! $cuota) {
                throw new RuntimeException('Este préstamo no tiene cuotas pendientes por pagar.');
            }

            $fechaPago = Carbon::parse($datos['fecha_pago']);
            $fechaEsperada = Carbon::parse($cuota->fecha_esperada);
            $diasMora = $fechaPago->gt($fechaEsperada) ? $fechaEsperada->diffInDays($fechaPago) : 0;

            $pendienteEnCuota = $this->pendienteEnCuota($cuota);
            $montoAbonado = round((float) $datos['monto_abonado'], 2);
            $pagosCreados = collect();

            if ($montoAbonado > $pendienteEnCuota) {
                $manejoExcedente = $datos['manejo_excedente'] ?? null;

                if (! in_array($manejoExcedente, ['abono_deuda', 'cobro_extra'], true)) {
                    throw new RuntimeException(
                        'El abono supera el monto pendiente de la cuota. Especifica "manejo_excedente" '.
                        '("abono_deuda" para restar el excedente de la deuda total, o "cobro_extra" si no debe afectar el préstamo).'
                    );
                }

                $excedente = round($montoAbonado - $pendienteEnCuota, 2);
                $cuota->estado = 'pagada';
                $cuota->save();

                // 'cobro_extra': el excedente queda registrado en monto_abonado de esta misma fila
                // pero no se aplica a la deuda (monto_aplicado se queda solo con lo que cubría la cuota).
                $montoAbonadoFila = $manejoExcedente === 'cobro_extra' ? $montoAbonado : $pendienteEnCuota;
                $pagosCreados->push($this->crearPago($prestamo, $cuota, $montoAbonadoFila, $pendienteEnCuota, $fechaPago, $diasMora));

                if ($manejoExcedente === 'abono_deuda') {
                    $pagosCreados = $pagosCreados->merge(
                        $this->aplicarExcedenteComoAbono($prestamo, $cuota, $excedente, $fechaPago)
                    );
                }
            } elseif ($montoAbonado < $pendienteEnCuota) {
                $faltante = round($pendienteEnCuota - $montoAbonado, 2);
                $this->aplicarPoliticaMora($prestamo, $cuota, $faltante, $diasMora);
                $pagosCreados->push($this->crearPago($prestamo, $cuota, $montoAbonado, $montoAbonado, $fechaPago, $diasMora));
            } else {
                $cuota->estado = 'pagada';
                $cuota->save();
                $pagosCreados->push($this->crearPago($prestamo, $cuota, $montoAbonado, $montoAbonado, $fechaPago, $diasMora));
            }

            $this->actualizarEstadoPrestamo($prestamo);

            return $pagosCreados;
        });
    }

    private function pendienteEnCuota(Cuota $cuota): float
    {
        $yaAplicado = (float) Pago::where('cuota_id', $cuota->id)->sum('monto_aplicado');

        return round((float) $cuota->monto_esperado - $yaAplicado, 2);
    }

    private function crearPago(
        Prestamo $prestamo,
        Cuota $cuota,
        float $montoAbonado,
        float $montoAplicado,
        Carbon $fechaPago,
        int $diasMora,
    ): Pago {
        $totalAplicadoPrevio = (float) Pago::where('prestamo_id', $prestamo->id)->sum('monto_aplicado');
        $saldoRestante = max(round($prestamo->monto_total - ($totalAplicadoPrevio + $montoAplicado), 2), 0.0);

        return Pago::create([
            'prestamo_id' => $prestamo->id,
            'cuota_id' => $cuota->id,
            'monto_abonado' => $montoAbonado,
            'monto_aplicado' => $montoAplicado,
            'fecha_pago' => $fechaPago->toDateString(),
            'dias_mora' => $diasMora,
            'saldo_restante_despues' => $saldoRestante,
        ]);
    }

    /**
     * Aplica un excedente de pago como abono a la deuda: lo va descontando de las
     * siguientes cuotas pendientes en orden, marcándolas como pagadas si alcanza,
     * dejando un registro de pago propio por cada cuota que resulte abonada.
     *
     * @return Collection<int, Pago>
     */
    private function aplicarExcedenteComoAbono(Prestamo $prestamo, Cuota $cuotaActual, float $excedente, Carbon $fechaPago): Collection
    {
        $pagosCreados = collect();

        $siguientes = $prestamo->cuotas()
            ->where('estado', '!=', 'pagada')
            ->where('numero_cuota', '>', $cuotaActual->numero_cuota)
            ->orderBy('numero_cuota')
            ->lockForUpdate()
            ->get();

        foreach ($siguientes as $siguiente) {
            if ($excedente <= 0) {
                break;
            }

            $pendiente = $this->pendienteEnCuota($siguiente);
            $aAplicar = min($excedente, $pendiente);

            if ($aAplicar >= $pendiente) {
                $siguiente->estado = 'pagada';
            }

            $siguiente->save();

            $pagosCreados->push($this->crearPago($prestamo, $siguiente, $aAplicar, $aAplicar, $fechaPago, 0));

            $excedente = round($excedente - $aAplicar, 2);
        }

        return $pagosCreados;
    }

    /**
     * Cuando el abono no alcanza a cubrir la cuota, aplica la política de mora
     * configurada en el préstamo (mantener | siguiente_pago | sumar_total).
     */
    private function aplicarPoliticaMora(Prestamo $prestamo, Cuota $cuotaActual, float $faltante, int $diasMora): void
    {
        $politica = $prestamo->politica_mora ?? 'mantener';

        if ($politica === 'siguiente_pago') {
            $siguiente = $prestamo->cuotas()
                ->where('estado', '!=', 'pagada')
                ->where('numero_cuota', '>', $cuotaActual->numero_cuota)
                ->orderBy('numero_cuota')
                ->lockForUpdate()
                ->first();

            if ($siguiente) {
                $siguiente->monto_esperado = round((float) $siguiente->monto_esperado + $faltante, 2);
                $siguiente->save();
                $cuotaActual->estado = 'pagada';
                $cuotaActual->save();

                return;
            }

            // No hay una siguiente cuota a la cual trasladar el faltante: se mantiene en la cuota actual.
            $politica = 'mantener';
        }

        if ($politica === 'sumar_total') {
            $pendientes = $prestamo->cuotas()
                ->where('estado', '!=', 'pagada')
                ->where('id', '!=', $cuotaActual->id)
                ->orderBy('numero_cuota')
                ->lockForUpdate()
                ->get();

            if ($pendientes->isNotEmpty()) {
                $porCuota = round($faltante / $pendientes->count(), 2);
                $acumulado = 0.0;
                $total = $pendientes->count();

                foreach ($pendientes as $index => $pendiente) {
                    $esUltima = $index === $total - 1;
                    $incremento = $esUltima ? round($faltante - $acumulado, 2) : $porCuota;
                    $acumulado += $incremento;

                    $pendiente->monto_esperado = round((float) $pendiente->monto_esperado + $incremento, 2);
                    $pendiente->save();
                }

                $cuotaActual->estado = 'pagada';
                $cuotaActual->save();

                return;
            }

            // No hay otras cuotas pendientes entre las cuales repartir el faltante: se mantiene en la cuota actual.
        }

        // Política 'mantener' (o sin alternativa disponible): la cuota queda abierta con el faltante pendiente.
        $cuotaActual->estado = $diasMora > 0 ? 'en_mora' : 'pendiente';
        $cuotaActual->save();
    }

    private function actualizarEstadoPrestamo(Prestamo $prestamo): void
    {
        $quedaPendiente = $prestamo->cuotas()->where('estado', '!=', 'pagada')->exists();

        if (! $quedaPendiente) {
            $estado = 'pagado';
        } elseif ($prestamo->cuotas()->where('estado', 'en_mora')->exists()) {
            $estado = 'en_mora';
        } else {
            $estado = 'activo';
        }

        if ($prestamo->estado !== $estado) {
            $prestamo->update(['estado' => $estado]);
        }
    }
}
